grai-integrations/source-bigquery/src/grai_source_bigquery/models.py

```diff
@@ -0,0 +1,173 @@
+from enum import Enum
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import BaseModel, Field, root_validator, validator
+
+
+class BigqueryNode(BaseModel):
+    pass
+
+
+class ID(BigqueryNode):
+    name: str
+    namespace: str
+    full_name: str
+
+    class Config:
+        extra = "forbid"
+
+
+class TableID(ID):
+    table_schema: str
+
+    @root_validator(pre=True)
+    def make_full_name(cls, values: Dict) -> Dict:
+        if values.get("full_name", None) is None:
+            values["full_name"] = f"{values['table_schema']}.{values['name']}"
+        return values
+
+
+class ColumnID(ID):
+    table_schema: str
+    table_name: str
+
+    @root_validator(pre=True)
+    def make_full_name(cls, values: Dict) -> Dict:
+        full_name = values.get("full_name", None)
+        if values.get("full_name", None) is None:
+            values["full_name"] = f"{values['table_schema']}.{values['table_name']}.{values['name']}"
+        return values
+
+    @validator("table_name")
+    def validate_name(cls, value):
+        if value.startswith('"') and value.endswith('"'):
+            return value
+        return value.lower()
+
+
+class Column(BigqueryNode):
+    name: str = Field(alias="column_name")
+    table: str
+    column_schema: str = Field(alias="schema")
+    data_type: str
+    is_nullable: bool
+    namespace: str
+    default_value: Any = Field(alias="column_default")
+    is_pk: Optional[bool] = False
+    full_name: Optional[str] = None
+
+    class Config:
+        allow_population_by_field_name = True
+
+    @validator("full_name", always=True)
+    def make_full_name(cls, full_name: Optional[str], values: Dict) -> str:
+        if full_name is not None:
+            return full_name
+        result = f"{values['column_schema']}.{values['table']}.{values['name']}"
+        return result
+
+    @validator("name")
+    def validate_name(cls, value):
+        if value.startswith('"') and value.endswith('"'):
+            return value
+        return value.lower()
+
+
+class Constraint(str, Enum):
+    foreign_key = "f"
+    primary_key = "p"
+    belongs_to = "bt"
+
+
+class Edge(BaseModel):
+    source: Union[ColumnID, TableID]
+    destination: Union[ColumnID, TableID]
+    definition: Optional[str]
+    constraint_type: Constraint
+    metadata: Optional[Dict] = None
+
+
+class TableType(str, Enum):
```

Comment:
We might have to update the table types for bigquery. From the [docs](https://cloud.google.com/bigquery/docs/information-schema-tables) it looks like they have a few different potential types.

```
The table type; one of the following:

    BASE TABLE: A standard [table](https://cloud.google.com/bigquery/docs/tables-intro)
    CLONE: A [table clone](https://cloud.google.com/bigquery/docs/table-clones-intro) ([Preview](https://cloud.google.com/products#product-launch-stages))
    SNAPSHOT: A [table snapshot](https://cloud.google.com/bigquery/docs/table-snapshots-intro)
    VIEW: A [view](https://cloud.google.com/bigquery/docs/views-intro)
    MATERIALIZED VIEW: A [materialized view](https://cloud.google.com/bigquery/docs/materialized-views-intro)
    EXTERNAL: A table that references an [external data source](https://cloud.google.com/bigquery/external-data-sources)
```